retrieval/scrape.go

```diff
@@ -446,6 +461,28 @@ func (sl *scrapeLoop) stop() {
 	<-sl.done
 }
 
+func (sl *scrapeLoop) processScrapeResult(samples model.Samples, scrapeErr error, start time.Time) error {
+	// Collect samples post-relabelling and label handling in a buffer.
+	buf := &bufferAppender{buffer: make(model.Samples, 0, len(samples))}
```

Comment:
I'm afraid we have already piled up way too many allocation-hostile code paths. And now we are using the badness to justify more badness.

We just had a measurement of a beefy server where simply putting query load onto it almost doubled the RAM usage. It needs more investigation, but it smells a lot like heap fragmentation.

The path added here will have vastly different allocation sizes, so it will probably create plenty of fragmentation.

It needs a lot more investigation for proof, but it's on the other hand fairly easy to short-cut the buffer in case it isn't needed.